RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,214 @@
+    int64_t n, // number of columns in A
+    int64_t d, // number of rows in sketch of A
+    int64_t k, // number of nonzeros in each column of the sketching operator
+    T *A, // buffer of size at least m*n.
+    int64_t lda, // leading dimension for mat(A).
+    T *V_sk, // buffer of size at least d*n.
+    T *sigma_sk, //buffer of size at least n.
+    RandBLAS::base::RNGState<RNG> state
+) {
+    T* buff_A = A;
+    T* buff_A_sk = V_sk;
+    if (m < n)
+        throw std::invalid_argument("Input matrix A must have at least as many rows as columns.");
+    
+    // step 1.1: define the sketching operator
+    RandBLAS::sparse::SparseDist D{
+        .n_rows = d,
+        .n_cols = m,
+        .vec_nnz = k
+    };
+    RandBLAS::sparse::SparseSkOp<T> S(D, state);
+    auto next_state = RandBLAS::sparse::fill_sparse(S);
+
+    // step 1.2: sketch the data matrix
+    int64_t lda_sk;
+    if (layout_A == blas::Layout::RowMajor) {
+        assert(lda >= n);
+        lda_sk = n;
+    } else {
+        assert(lda >= m);
+        lda_sk = d;
+    }
+    RandBLAS::util::safe_scal(d*n, 0.0, buff_A_sk, 1);
+    RandBLAS::sparse::lskges<T>(
+        layout_A,
+        blas::Op::NoTrans,
+        blas::Op::NoTrans,
+        d, n, m,
+        1.0, S, 0, 0,
+        buff_A, lda,
+        0.0, buff_A_sk, lda_sk
+    );
+
+    // step 2: compute SVD of sketch
+    T* ignore = nullptr;
+    if (layout_A == blas::Layout::RowMajor) {
+        //
+        //      buff_A_sk is stored in row-major format; we want its singular values
+        //      and right singular vectors.
+        //
+        //      We get around LAPACK's column-major restriction by using the fact that
+        //          memory(A_sk, RowMajor) == memory(transpose(A_sk), ColumnMajor).
+        //      So we'll tell LAPACK that we want the singular values and left
+        //      singular vectors of tranpose(A_sk).
+        //
+        lapack::Job jobu = lapack::Job::OverwriteVec;
+        lapack::Job jobvt = lapack::Job::NoVec;
+        lapack::gesvd(jobu, jobvt, n, d, buff_A_sk, n, sigma_sk, ignore, 1, ignore, 1);
+        // interpret buff_A_sk in row-major
+    } else {
+        lapack::Job jobu = lapack::Job::NoVec;
+        lapack::Job jobvt = lapack::Job::SomeVec;
+        T *Vt = new T[n * n]{};
+        lapack::gesvd(jobu, jobvt, d, n, buff_A_sk, d, sigma_sk, ignore, 1, Vt, n);
+        for (int i = 0; i < n; ++i) {
+            // buff_A_sk is now just scratch space that points to V_sk.
+            // We just computed its SVD as a d-by-n matrix in column-major order.
+            // Now we want to write to the first n^2 elements of buff_A_sk with
+            // n "n-vectors" in column-major order.
+            blas::copy(n, &Vt[i], n,  &buff_A_sk[i*n], 1);
+        }
+        delete[] Vt;
+    }
+    return next_state;
+}
+
+
+/**
+ * Accepts the right singular vectors and singular values of some tall
+ * matrix "T", along with a regularization parameter mu.
```

Comment:
I'm not a huge fan of the matrix T since T is a type variable for this function and is not any of the parameters, so it might be a bit confusing

Reply:
Good catch!